### DISCUSSION TITLE
Release v0.1.8-preview

### DISCUSSION THREAD

**Discussion by jason-dou** [microsoftgraph/msgraph-bicep-types#179] [Announcements]
We're excited to announce the release of v0.1.8-preview! This version is exclusively available on the [Microsoft Artifact Registry (MAR)](https://mcr.microsoft.com/) and can be accessed using dynamic types. Please note that it requires a minimum Bicep version of [0.30.3](https://github.com/Azure/bicep/releases/tag/v0.30.3). For more information on how to migrate to dynamic types, check out [How to migrate to dynamic types](https://aka.ms/graphBicepDynamicTypes). [Quickstart templates](https://github.com/microsoftgraph/msgraph-bicep-types/tree/main/quickstart-templates) are also updated with examples using Graph Bicep resources in v0.1.8-preview.

> [!IMPORTANT]
> Built-in types are deprecated and will be retired on January 24, 2025. Until the retirement date, built-in types, denoted by `extension microsoftGraph`, will coexist with the new dynamic types. Any Microsoft Graph Bicep type changes will only be available through new versions of the dynamic types.

## What's changes
- Added `isManagementRestricted` property in `groups@v1.0` resource type

## Issues fixed
- Fixed an issue with managing app role assignment for all principal types #161 

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoftgraph/msgraph-bicep-types/releases/tag/v0.1.8-preview'>Release v0.1.8-preview</a>.</em>

**Comment by slavizh** [comment]
@jason-dou is there a way of the version of the type inside the .bicep file? Being defined in bicepconfig.json and referenced in .bicep file could lead the to problem where end user can take the .bicep file modify its bicepconfig.json with incorrect version that was used to verify the .bicep template functionality. It really looks like the version of the extension should not be defined in bicepconfig.json .

**Reply by anthony-c-martin** [reply]
You can also define the version of the type inside the `.bicep` file - it's just more verbose.

For example:
```bicep
extension microsoftGraphV1_0
```
With bicepconfig
```json
{  
    "experimentalFeaturesEnabled": {  
        "extensibility": true  
    },
    "extensions": {
        "microsoftGraphV1_0": "br:mcr.microsoft.com/bicep/extensions/microsoftgraph/v1.0:0.1.8-preview"   
    }
}
```

Could instead be written as the following:
```bicep
extension 'br:mcr.microsoft.com/bicep/extensions/microsoftgraph/v1.0:0.1.8-preview' as microsoftGraphV1_0
```
With bicepconfig
```json
{  
    "experimentalFeaturesEnabled": {  
        "extensibility": true  
    }
}
```

**Reply by slavizh** [reply]
FYI may be the path should have been microsoft-graph instead of microsoftgraph. It is annoying if something is not camelCase or some other format that is not readable.

**Reply by slavizh** [reply]
Also note that microsoftGraphV1_0 cannot be used. Gives error: 'This import statement has an invalid alias value. The import alias 'microsoftGraphV1_0' must be 1 or more characters, starting with a letter, and only consisting of letters or digits. Please see https://aka.ms/arm-syntax for usage details.'.
It is a little bit restrictive.

**Reply by dkershaw10** [reply]
Just seeing this.  @slavizh are you still seeing this issue?
I cannot repro your alias naming problem when using bicepconfig OR defining the version directly in the bicep file.

**Reply by slavizh** [reply]
@dkershaw10 yes, you need to start deployment to see it. I am using Az PowerShell for deployment if that matters at all.

![image](https://github.com/user-attachments/assets/41d4819f-43e2-430c-8c16-80476ea3bd07)

**Reply by dkershaw10** [reply]
@slavizh I've had trouble in the past with Az PowerShell config not picking up the latest Bicep versions.  That *could* be the source of your problems. https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/install#install-manually.  It's one reason that I use Azure CLI, as it tells you when there's a Bicep update and it's easier to update.

**Reply by slavizh** [reply]
@dkershaw10 I doubt. I am running this on DevOps hosted agents which are used for others Bicep deployments and they do not have problems. Also this works as soon as I correct the name to not have '_'. Additionally the above is ARM error not Bicep CLI version. And as last output shows which bicep cli version is used to compile the template before deployment. The above is clearly general extensions error which @anthony-c-martin should look at it.

![image](https://github.com/user-attachments/assets/dc906e91-5826-4b48-b8e9-23e0b19d53b6)

**Reply by dkershaw10** [reply]
Fair enough - I thought it was a long-shot given your long history with ARM, Bicep and Azure :).
@anthony-c-martin any ideas?  I've not been able to repro this issue.

**Reply by anthony-c-martin** [reply]
@slavizh is there any way you'd be able to share code samples (or a minimal repro if possible)? Could you confirm the version of the extension you have imported?

What's especially puzzling to me is that the name of the import should be set by the extension itself, rather than by anything the user types. In my case it's set to "MicrosoftGraph", even though I've imported it as "microsoftGraphV1_0" in Bicep:

![image](https://github.com/user-attachments/assets/908647d4-af5f-4b13-9240-b720cb8a9c55)

**Reply by anthony-c-martin** [reply]
Ah I think I can repro if I alias it myself with `as microsoftGraphV1_0` in the `extension` declaration. Is that what you're doing?

![image](https://github.com/user-attachments/assets/fb9b90f4-a71b-4909-944b-db5dd62b44dc)

This definitely looks like a bug in Bicep.

**Reply by slavizh** [reply]
@anthony-c-martin yes. Do you want me to open issue for it? It would be good if the name supports _ or some type of other character like - or . but if it is not possible at least it should be flagged by VSC before deployment.

**Reply by anthony-c-martin** [reply]
I've submitted a PR to have the backend service match Bicep's more permissive naming validation. This'll take a long time to get deployed to the live service (ETA - probably at least 3 weeks). In the meantime, the workaround would be to avoid using an alias that contains "_" characters.

**Reply by slavizh** [reply]
@anthony-c-martin great. Thanks! 3 weeks is fine.

**Reply by anthony-c-martin** [reply]
Just to set expectations, there is now a pretty extensive change freeze in effect for Azure services over the winter holiday period. We have a window at the beginning of December; I hope this'll be deployed then, but can't make any guarantees in case something unexpected happens. If we miss that window, then we're looking at starting the deployment in early January.

**Reply by dkershaw10** [reply]
Thanks for the heads up.  I'll update all usage examples not to use the "_" character.

**Reply by slavizh** [reply]
that is fine.

**Reply by dkershaw10** [reply]
FYI - doc PR is in progress and should be published some time this week.

**Comment by dkershaw10** [comment]
@slavizh we can add doc examples of defining the version of the type inside the .bicep file too.

**Reply by slavizh** [reply]
@dkershaw10 awesome. May be advantages and disadvantages to both approaches would be helpful for some folks.

**Reply by dkershaw10** [reply]
@slavizh Please see updated topic here:  https://learn.microsoft.com/graph/templates/how-to-dynamic-types. Hopefully this addresses your concerns.

**Comment by dkershaw10** [comment]
@anthony-c-martin @shenglong I was creating some new docs on dynamic types, per the conversation above and I have a question, because I don't have a way to test the scenario.  cc: @jason-dou 

So say I do this:
```Bicep
extension 'br:mcr.microsoft.com/bicep/extensions/microsoftgraph/v1.0:<v1.0-version>' as graphV1_0
extension 'br:mcr.microsoft.com/bicep/extensions/microsoftgraph/beta:<beta-version>' as graphBeta
```

Other than generating the ARM template with the alias names, I don't see any real benefit to the `as` keyword for this scenario.  The only scenario I could think of that *might* be valuable is differentiating between two type versions in Microsoft Graph v1.0 for example, like:

```Bicep
extension 'br:mcr.microsoft.com/bicep/extensions/microsoftgraph/v1.0:<v1.0-version-1>' as graphV1_0_latest
extension 'br:mcr.microsoft.com/bicep/extensions/microsoftgraph/v1.0:<v1.0-version-2>' as graphV1_0_older

// using latest
resource group 'graphV1_0_latest:Microsoft.Graph/groups@v1.0' existing = {
    uniqueName: groupName
}

// using an older version
resource app 'graphV1_0_older:Microsoft.Graph/applications@v1.0' existing = {
    uniqueName: appName
}
```

I can't fully test this because we only have the one type version for v1.0 at the moment.
**Do you think that this should work?**

Also IntelliSense isn't working properly when you try to use the fully-qualified type name.

**Reply by anthony-c-martin** [reply]
> Other than generating the ARM template with the alias names, I don't see any real benefit to the as keyword for this scenario. The only scenario I could think of that might be valuable is differentiating between two type versions in Microsoft Graph v1.0 for example, like:

I think your analysis is correct here - it's there for completeness, but isn't something I'd expect to see much outside of some very niche use-cases.
1. Exactly the scenario you've called out - if trying to disambiguate between a type of the same name that's available in 2 different imported providers.
1. If the user wants to define symbols and avoid collisions. The graph extension has a pretty unique name, so probably less likely to run into here. But for example, if the user is using the `az` provider, and wanted to define a variable named `az` - they would have to alias the namespace import.
1. (really an extension of 2). Namespaces can declare functions (although not used by msgraph), so if for example the msgraph provider ever had a need for a function like `lookupUser(<upn>)`, this would also be useful to disambiguate so you could do e.g. `graphV1_0.lookupUser(...)`. This extends beyond wanting to import graph twice - it's conceivable that another namespace (e.g. `az`) could define a function of the same name.

**Comment by dkershaw10** [comment]
Howdy folks,

Just an update here on dynamic types.  Here and in our [public docs](https://learn.microsoft.com/graph/templates/how-to-dynamic-types?view=graph-bicep-1.0#migrate-to-using-dynamic-types-from-built-in-types), we've asked folks to migrate to dynamic types (from using the built-in types).  Additionally, there are warnings in both VS/VS Code and when deploying a template file, if you are still using built-in types.

The notices and warning indicate that built-in types will no longer work from January 24.  However, we're pushing this date out by a little over a month (likely first week in March, or whenever the monthly Bicep release happens in March). We're doing this because we want to give you more time and we discovered some issues with dynamic types in some of our air-gapped clouds, as part of our final testing of national cloud support (which will be available soon).
